pkg/storage/action_cache_asset_store.go

```diff
@@ -261,9 +248,15 @@ func (rs *actionCacheAssetStore) Put(ctx context.Context, ref *asset.AssetRefere
 		},
 	}
 
-	// Check if the input asset is a directory or blob
-	d, err := rs.assetToDirectory(ctx, data, instance)
-	if err == nil {
+	if data.Type == asset.Asset_DIRECTORY {
+		d, err := rs.assetToDirectory(ctx, data, instance)
+		if err != nil {
+			// Users will hit this if they upload an digest referencing an
+			// arbitary Blob in `PushDirectory` or a digest that does not
+			// reference any blob at all.
+			return fmt.Errorf("digest in directory asset does not reference a Directory: %v", err)
```

Comment:
Have you made sure we actually hit this in this case?  In the original issue, `assetToDirectory` silently unmarshals stuff into bad `Directory` objects without failing, so I have a suspicion that this will still have the same issue.

Reply:
- If the `PushDirectoryRequest` contains a `root_directory_digest` referencing a blob of random data you will hit this case ("digest in directory asset does not reference a Directory").
 - If  the `PushDirectoryRequest` contains a `root_directory_digest` referencing something that looks like a Directory but isn't fully valid you get a different error. For my test case that error message looks like "unknown digest function" but a different looks-like-but-is-not-directory would give a different error.

Reply:
I have also just realised that I should be using `status.Errorf` rather than `fmt.Errorf`. Fixed this in a commit

Reply:
Great, that seems like the behaviour we're after.